File: javascript/leetcode/1110. Delete Nodes And Return Forest.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number[]} to_delete
 * @return {TreeNode[]}
 */
var delNodes = function (root, to_delete) {
  const heads = [];
  const seen = new Set(to_delete);

  dfs(root, heads, seen, true);

  return heads;
};

function dfs(root, heads, seen, is_root) {
  if (!root) return null;

  const deleted = seen.has(root.val);

  if (is_root && !deleted) heads.push(root);

  root.left = dfs(root.left, heads, seen, deleted);
  root.right = dfs(root.right, heads, seen, deleted);

  return deleted ? null : root;
}

File: c++/leetcode/1110. Delete Nodes And Return Forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  TreeNode *del(TreeNode *, vector<TreeNode *> &, const unordered_set<int> &, const bool &);
  vector<TreeNode *> delNodes(TreeNode *root, vector<int> &to_delete)
  {
    vector<TreeNode *> heads;
    unordered_set<int> seen{to_delete.begin(), to_delete.end()};

    del(root, heads, seen, true);

    return heads;
  }
};

TreeNode *Solution::del(TreeNode *root, vector<TreeNode *> &heads, const unordered_set<int> &seen, const bool &is_root)
{
  if (!root)
    return nullptr;

  auto deleted = seen.count(root->val);

  if (is_root && !deleted)
    heads.push_back(root);

  root->left = del(root->left, heads, seen, deleted);
  root->right = del(root->right, heads, seen, deleted);

  return deleted ? nullptr : root;
}